master/internal/api_trials.go

```diff
@@ -77,3 +81,59 @@ func (a *apiServer) TrialLogs(
 		time.Sleep(batchWaitTime)
 	}
 }
+
+func (a *apiServer) GetTrialCheckpoints(
+	_ context.Context, req *apiv1.GetTrialCheckpointsRequest,
+) (*apiv1.GetTrialCheckpointsResponse, error) {
+	_, _, err := trialStatus(a.m.db, req.Id)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := &apiv1.GetTrialCheckpointsResponse{}
+	resp.Checkpoints = []*checkpointv1.Checkpoint{}
+
+	switch err := a.m.db.QueryProto("get_checkpoints_for_trial", &resp.Checkpoints, req.Id); {
+	case err == db.ErrNotFound:
+		return nil, status.Errorf(
+			codes.NotFound, "no checkpoints found for trial %d", req.Id)
+	case err != nil:
+		return nil,
+			errors.Wrapf(err, "error fetching checkpoints for trial %d from database", req.Id)
+	}
+
+	a.filter(&resp.Checkpoints, func(i int) bool {
+		v := resp.Checkpoints[i]
+
+		found := false
+		for _, state := range req.States {
+			if state == v.State {
+				found = true
+				break
+			}
+		}
+
+		if len(req.States) != 0 && !found {
+			return false
+		}
```

Comment:
it'd be nice if this logic could be more modular. I think it's not the first time I've seen it.

Reply:
Yea I agree. I worry that we won't be able to capture an overall state filter without reflect though. For example the checkpoint states are slightly different that experiment states and have a different protobuf type.

If you feel strongly I can open a PR to unify the checkpoint logic across trial and experiment endpoints.